### PR TITLE
Google auth

### PR DESCRIPTION
**Pull request by shashankkarthik**
# Description

- Added Google Sign-In button to sign in page.
- Google Sign-In for existing users by matching emails.
- Google Sign-in for nonexistent users by creating new user from Google User data.

## Checklist

- [ ] This PR can be reviewed in under 30 minutes
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have assigned reviewers to this PR.

### CODE REVIEW

File: packages/client/package.json [hunk 1]
```diff
@@ -26,12 +26,14 @@
     "yup": "^0.32.11"
   },
   "devDependencies": {
+    "@react-oauth/google": "*",
```
Comment (cbolles) [inline packages/client/package.json]: Should this be a regular dependency?

File: packages/client/src/main.tsx [hunk 2]
```diff
@@ -1,9 +1,12 @@
 import React from 'react';
 import ReactDOM from 'react-dom/client';
 import { App } from './app';
+import { GoogleOAuthProvider } from '@react-oauth/google';
 
 ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
-  <React.StrictMode>
-    <App />
-  </React.StrictMode>
+  <GoogleOAuthProvider clientId="920265659839-vu93dpp83vk1gaotvq5psjeeobjkbmlt.apps.googleusercontent.com">
```
Comment (cbolles) [inline packages/client/src/main.tsx]: That client ID looks like it should probably be an environment variable

**Review by cbolles — CHANGES_REQUESTED** [review]
Looks good, only a couple of minor suggestions on my end

File: packages/client/src/main.tsx (hunk 2 above)
Comment (gregfrasco) [inline packages/client/src/main.tsx]: We should also put the provider inside of React.StrictMode and not above it, unless there is a reason to not be in strict mode

File: packages/client/src/pages/sign-in.tsx [hunk 3]
```diff
@@ -53,38 +61,46 @@ export const SignIn = () => {
...
-                <TextInput autoFocus fullWidth name="email" label="Email Address" type="email" autoComplete="email" margin="normal" required />
-                <PasswordInput name="password" label="Password" fullWidth autoComplete="current-password" required margin="normal" />
-              </CardContent>
-            </Card>
-            <SubmitButton fullWidth variant="contained" color="primary" sx={{ my: 2 }}>
-              Sign In
-            </SubmitButton>
-          </Form>
-        </Formik>
+        {project?.authMethods.emailAuth ? (
```
Comment (gregfrasco) [inline packages/client/src/pages/sign-in.tsx]: If you are conditionally rendering something where on condition renders null, use can just use the && operator
```suggestion
        {project?.authMethods.emailAuth && (
```
And then just close it
```
)}
```

File: packages/client/src/pages/sign-in.tsx [hunk 4]
```diff
@@ -53,38 +61,46 @@ export const SignIn = () => {
...
-          <Grid item xs>
-            <Button
-              onClick={() => {
-                navigate(Paths.FORGOT_PASSWORD);
-              }}
-            >
-              <Typography variant="body2">Forgot password?</Typography>
-            </Button>
-          </Grid>
+          {project?.authMethods.emailAuth ? (
```
Comment (gregfrasco) [inline packages/client/src/pages/sign-in.tsx]: You probably could include this in the emailAuth condition if you add a fragment to wrap both the form and the buttons.

File: packages/client/src/pages/sign-in.tsx [hunk 5]
```diff
@@ -99,6 +115,22 @@ export const SignIn = () => {
             <></>
           )}
         </Grid>
+        <Grid container>
```
Comment (gregfrasco) [inline packages/client/src/pages/sign-in.tsx]: If this is a single item you don't need a grid, you could just use a box and add styling or a stack to help with layout

File: packages/client/src/pages/sign-in.tsx [hunk 6]
```diff
@@ -99,6 +115,22 @@ export const SignIn = () => {
...
+          {project?.authMethods.googleAuth ? (
+            <Grid item>
+              <GoogleLogin
+                onSuccess={({ credential }) => {
+                  loginGoogle({ variables: { projectId: project?.id || '', credential: credential as string } }).catch((err) => {
+                    console.log(err);
+                    pushMessage('Invalid Credentials');
```
Comment (gregfrasco) [inline packages/client/src/pages/sign-in.tsx]: This message should be more helpful to users
```suggestion
                    pushMessage('Unable to Authenticate with Google');
```

File: packages/server/src/auth/auth.service.ts [hunk 7]
```diff
@@ -9,6 +9,9 @@ import { UpdateStatus } from '../user/types/user.types';
 import { ConfigService } from '@nestjs/config';
 import { ProjectService } from '../project/project.service';
 import { NotificationService } from '../notification/notification.service';
+import { decode } from 'jsonwebtoken';
```
Comment (gregfrasco) [inline packages/server/src/auth/auth.service.ts]: Not used

File: packages/server/src/auth/auth.service.ts [hunk 8]
```diff
@@ -9,6 +9,9 @@ import { UpdateStatus } from '../user/types/user.types';
...
+import { HttpService } from '@nestjs/axios';
+import { last, lastValueFrom } from 'rxjs';
```
Comment (gregfrasco) [inline packages/server/src/auth/auth.service.ts]: last not used

**Review by gregfrasco — CHANGES_REQUESTED** [review]
Noce job, some minor changes

**Review by cbolles — DISMISSED** [review]

File: packages/server/src/auth/auth.service.ts [hunk 9]
```diff
@@ -67,6 +70,39 @@ export class AuthService {
...
+  /**
+   *
+   * @param projectId
+   * @param credential
+   * @returns JWT or 401 status code
+   */
+  async validateGoogle(projectId: string, credential: string): Promise<any> {
+    if (projectId == null || credential == null) {
+      throw new HttpException('Bad request: project id and credential required.', HttpStatus.BAD_REQUEST);
+    }
```
Comment (gregfrasco) [inline packages/server/src/auth/auth.service.ts]: Minor: you have validation thru decorators, so you shouldn't need to have it hear as well

**Review by gregfrasco — APPROVED** [review]

**Pull request merged**